File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
def solution(string):
    answer = True
    stack = []

    for char in string:
        if char == '(':
            stack.append(char)

        elif char == ')':
            if stack and stack[-1] == '(':
                stack.pop()
                continue
            stack.append(char)

    if stack:
        answer = False

    return answer

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;

class Solution {
    boolean solution(String s) {
        boolean answer = true;
        
        // Stack 과 Queue 둘다 Deque를 사용한다.
        Deque<Character> stack = new ArrayDeque<>();
        
        for(char c : s.toCharArray()){
            if(stack.isEmpty()){
                if (c == ')'){
                    return false;
                }
            } 
             
            if(c == '('){
                stack.push(c);
            } else {
                stack.pop();
            }
            
        }
        
        if(stack.isEmpty()){
            return true;
        }

        return false;
    }
}